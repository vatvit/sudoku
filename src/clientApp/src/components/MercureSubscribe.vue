<script>
  import './MercureMessageItem.vue';
  import MercureMessageItem from "@/components/MercureMessageItem.vue";

  let mercureSubscription;

  export default {
    props: {
      messages: {
        type: Array,
        default: [],
      },
      subscribedTopic: String,
    },
    data() {
      return {
        messages: [],
        subscribedTopic: '',
      };
    },
    components: {
      MercureMessageItem
    },
    mounted() {
      this.subscribe();
    },
    methods: {
      subscribe() {
        this.subscribedTopic = this.$refs.subscriptionTopicName.value;

        const url = new URL(window.sudoku.config.mercurePublicUrl);
        url.searchParams.append('topic', this.subscribedTopic);

        if (mercureSubscription) {
          mercureSubscription.close();
        }
        mercureSubscription = new EventSource(url);
        mercureSubscription.onmessage = this.onMessage;
      },
      unsubscribe() {
        this.subscribedTopic = '';
        mercureSubscription.close();
        mercureSubscription = undefined;
      },
      clearLog() {
        this.messages = [];
      },
      onMessage(message) {
        console.log(message);
        this.messages.unshift({
          id: message.lastEventId,
          data: message.data,
        });
        console.log(this.messages);
      },
    },
  }
</script>

<template>
  <div id="status">Subscribed topic: <b>{{ subscribedTopic || 'no subscription' }}</b></div>
  <input id="subscriptionTopicName" ref="subscriptionTopicName" name="topic" placeholder="topic name e.g. 'my-private-topic'" value="*">
  <button id="subscribe" @click="subscribe">Subscribe</button>
  <button id="unsubscribe" @click="unsubscribe">Unsubscribe</button>
  <button id="clear" @click="clearLog">Clear log</button>
  <div id="messages" style="border: 1px red; padding: 5px;">
    Messages:
    <MercureMessageItem v-for="message in messages" :key="message.id" :id="message.id" :data="message.data" />
  </div>
</template>
