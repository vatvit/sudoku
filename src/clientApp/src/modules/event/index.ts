export const buildEventSource = (url: URL | string) => {
    return new EventSource(url, {});
};