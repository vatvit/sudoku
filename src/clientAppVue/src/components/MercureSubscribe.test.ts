import { mount } from '@vue/test-utils'
import MercureSubscribe from './MercureSubscribe.vue'
// @ts-ignore
import * as eventsourcemock from 'eventsourcemock'
import * as event from '../modules/event'

const EventSource = eventsourcemock.default
const sources = eventsourcemock.sources

const buildEventSourceSpy = vi.spyOn(event, 'buildEventSource')

const messageEvent = new MessageEvent('foo', {
    data: '1',
});

const mercurePublicUrl = 'http://testMercurePublicUrl.example.com'

test('mount component', async () => {

    expect(MercureSubscribe).toBeTruthy()

    buildEventSourceSpy.mockReturnValue(new EventSource(mercurePublicUrl))

    const wrapper = mount(MercureSubscribe, {
        global: {
            provide: {
                $sudoku: {config: {mercurePublicUrl: mercurePublicUrl}}
            }
        }
    })

    sources[mercurePublicUrl].emit(
        messageEvent.type,
        messageEvent
    );

    expect(wrapper.text()).toContain('Subscribed topic: no subscription')
    expect(wrapper.text()).toMatchSnapshot()
})
