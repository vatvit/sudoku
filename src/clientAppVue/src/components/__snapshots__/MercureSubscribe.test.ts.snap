// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `"Subscribed topic: no subscriptionSubscribeUnsubscribeClear logpublish Messages:"`;
